src/main/java/org/mqttbee/mqtt/datatypes/MqttUtf8StringImpl.java

```diff
@@ -64,43 +65,78 @@
 
     /**
      * Validates and decodes a UTF-8 encoded String from the given byte array.
+     * <p>
+     * Note: the given byte array must not be longer than {@link MqttBinaryData#MAX_LENGTH}.
      *
      * @param binary the byte array with the UTF-8 encoded data to decode from.
      * @return the created UTF-8 encoded String or null if the byte array does not contain a well-formed UTF-8 encoded
      * String.
      */
     @Nullable
-    public static MqttUTF8StringImpl from(@NotNull final byte[] binary) {
-        return containsMustNotCharacters(binary) ? null : new MqttUTF8StringImpl(binary);
+    public static MqttUtf8StringImpl from(@NotNull final byte[] binary) {
+        return (binary.length > MqttBinaryData.MAX_LENGTH || containsMustNotCharacters(binary)) ? null :
+                new MqttUtf8StringImpl(binary);
     }
 
     /**
      * Validates and creates a UTF-8 encoded String from the given string.
+     * <p>
+     * Note: The given string
+     * <ul>encoded in UTF-8 must not be longer than {@link MqttBinaryData#MAX_LENGTH}</ul>
+     * <ul>and must not contain the null character U+0000 and UTF-16 surrogates, as these are forbidden according to
+     * the MQTT 5 Specification.</ul>
      *
      * @param string the UTF-16 encoded Java string.
      * @return the created UTF-8 encoded String or null if the string is not a valid UTF-8 encoded String.
+     * @throws IllegalArgumentException if the given string encoded in UTF-8 is longer than {@link
+     *                                  MqttBinaryData#MAX_LENGTH} or contains forbidden characters.
      */
-    @Nullable
-    public static MqttUTF8StringImpl from(@NotNull final String string) {
-        return containsMustNotCharacters(string) ? null : new MqttUTF8StringImpl(string);
+    @NotNull
+    public static MqttUtf8StringImpl from(@NotNull final String string) {
+        // Perform a quick check on string length, since calculating the exact number of bytes used in UTF-8 encoding
+        // might be an expensive operation.
+        // Note: Java strings are represented in UTF-16 (fixed-width 16-bit chars for code points U+0000 to U+FFFF).
+        //       Supplementary characters (code points greater than U+FFFF) are represented as a pair of char values
+        //       from the high-surrogate and low-surrogate ranges, hence their representation needs 4 bytes.
+        //       The UTF-8 representation uses a variable number of 1 to 3 bytes to represent code points from U+0000 to
+        //       U+FFFF. Representation of supplementary characters in UTF-8 also needs 4 bytes.
+        //       String.length() returns the number of 16-bit char values.
+        //       Hence, string.length() * 3 is an upper bound of the number of bytes needed to represent a Java string
+        //       (UTF-16) in UTF-8.
+        if (string.length() * 3 > MqttBinaryData.MAX_LENGTH) {
```

Comment:
The check should be extracted to `static boolean isInRange(String)` beneath `containsMustNotCharacters(String)`

Reply:
A boolean return value does not really make sense, since in the false case the method will throw. So I would suggest a `static void checkStringLength(String)`.